model/map-jpa/src/main/resources/META-INF/users/jpa-users-changelog-1.xml

```diff
@@ -0,0 +1,170 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+Copyright 2022 Red Hat, Inc. and/or its affiliates
+and other contributors as indicated by the @author tags.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+-->
+
+
+<databaseChangeLog xmlns="http://www.liquibase.org/xml/ns/dbchangelog" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+                   xmlns:ext="http://www.liquibase.org/xml/ns/dbchangelog-ext"
+                   xsi:schemaLocation="http://www.liquibase.org/xml/ns/dbchangelog http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-3.1.xsd
+                    http://www.liquibase.org/xml/ns/dbchangelog-ext http://www.liquibase.org/xml/ns/dbchangelog/dbchangelog-ext.xsd">
+
+    <changeSet author="keycloak" id="users-1">
+
+        <createTable tableName="kc_user">
+            <column name="id" type="UUID">
+                <constraints primaryKey="true" nullable="false"/>
+            </column>
+            <column name="version" type="INTEGER" defaultValueNumeric="0">
+                <constraints nullable="false"/>
+            </column>
+            <column name="metadata" type="json"/>
+        </createTable>
+        <ext:addGeneratedColumn tableName="kc_user">
+            <ext:column name="entityversion" type="INTEGER" jsonColumn="metadata" jsonProperty="entityVersion"/>
+            <ext:column name="realmid" type="KC_KEY" jsonColumn="metadata" jsonProperty="fRealmId"/>
+            <ext:column name="username" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fUsername"/>
+            <ext:column name="firstname" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fFirstName"/>
+            <ext:column name="lastname" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fLastName"/>
+            <ext:column name="email" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fEmail"/>
+            <ext:column name="federationlink" type="VARCHAR(255)" jsonColumn="metadata" jsonProperty="fFederationLink"/>
+            <ext:column name="enabled" type="BOOLEAN" jsonColumn="metadata" jsonProperty="fEnabled"/>
+            <ext:column name="emailverified" type="BOOLEAN" jsonColumn="metadata" jsonProperty="fEmailVerified"/>
+            <ext:column name="timestamp" type="BIGINT" jsonColumn="metadata" jsonProperty="fCreatedTimestamp"/>
+        </ext:addGeneratedColumn>
+        <createIndex tableName="kc_user" indexName="user_entityVersion">
+            <column name="entityversion"/>
+        </createIndex>
```

Comment:
I think this is missing a unique index for username + realmid, as a username should exist only once per realm

Reply:
Another thought: as the two missing indexes haven been found by tests, there should be two new tests for unique email in realm and unique user name in realm.

Reply:
... but these new tests should be in a new issue and pull request.

Reply:
Agreed. If that is in place, then having an index targeting only the `realmid` might not be necessary, right?

Reply:
It would not be necessary if the realmid is the first column of that index. If it is the second one, it will still be necessary AFAIK. I was arguing above that an index with userid first might be more efficient for the constraint, as a lot of usernames would share the same realmid, but not vice versa. A load test would still need to prove that. You can decide to defer sich optimizations.

Reply:
On the missing tests, I've checked `UserTest` and it does have tests for the unique username and unique emails. But these are enforced at the provider level, so it never reaches the DB where we would notice the missing constraints. Do you think we should add tests that work at the entity level directly (bypassing the provider) to ensure the constraints are in place?

Reply:
Thanks for looking that up. I see that this happens for example in MapUserAdapter for those stores that don't implement that constraint on the store level. This will be "best effort". Hm. It would be great to have an isolated test on the JPA map store, which we don't have and probably never will have. So let's skip the automated test, I'll do a manual test once the index is there by trying to update the DB. 

So it is only to add the index.

Reply:
> Another thought: as the two missing indexes haven been found by tests, there should be two new tests for unique email in realm and unique user name in realm.

Note to my future self: before writing something like this, have a look at the existing code.